base/db/R/dbfiles.R

```diff
@@ -636,3 +636,63 @@ dbfile.id <- function(type, file, con, hostname=PEcAn.remote::fqdn()) {
     invisible(NA)
   }
 }
+
+#' Function to Insert Records into the Formats and Variables Tables
+#' @name dbfile.format.variable.insert
+#' @param format_name Format Name
+#' @param header TRUE/FALSE Boolean describing whether or not the file contains a header
+#' @param skip The number of lines in the header
+#' @param mimetype_id The mimetype id of the mimetype that is associated with the format
+#' @param format_notes Any useful notes that describe the format being created
+#' @param con Connection object to connect to bety
+#' @param format_variables A TRUE/FALSE Boolean used to toggle a format-variables db query
+#' @param description variable descripthon
+#' @param units variable units
+#' @param var_notes notes to further describe the variable
+#' @param var_name variable name
+#' @param max maximum possible value
+#' @param min minimum possible value
+#' @param standard_name PEcAn specific name? 
+#' @param label ?
+#' @param type ?
+#' @author Liam Burke
+#'
+#' @return formatid
+#' @export
+#' 
+#' @examples
+#' \dontrun{
+#' bety <- betyConnect()
+#'   dbfile.format.variable.insert('Test_Format', header = TRUE, skip = 2, mimetype_id = 1060, con = bety$con, format_variables = FALSE)
+#' }
+dbfile.format.variable.insert <- function(format_name, header, skip, mimetype_id, format_notes = "", con, format_variables = FALSE, 
+                                          description = NA, units = NA, var_notes = NA, var_name = NA, max = NA, min = NA, standard_name = NA, label = NA, type = NA){
+ 
+  formatid <- get.id(
+    table = "formats",
+    colnames = c('mimetype_id', 'name'),
+    values = c(mimetype_id, format_name),
+    con = con,
+    create = TRUE,
+    dates = TRUE)
+
+    #### Formats Query ####
+    cmd <- paste0("INSERT INTO formats ",
+                   "(header, skip, mimetype_id, notes, name) VALUES (",
+                   header, ", ", skip, ", '", mimetype_id, "', '", format_notes, "','", format_name, "')")
+
+    db.query(query = cmd, con = con)
```

Comment:
@infotroph I see you point and agree that the use of raw SQL is potentially dangerous. `DBI::dbWriteTable` seems like it is a better solution in most instances. That said, it appears that this function only returns TRUE/FALSE and you can't specify the "RETURNING" argument which I've asked @LiamBurke24 to add. I find this to be an unfortunate and odd design limitation.

Reply:
@mdietze I think I follow, but to be clear: "this function" being `DBI::dbWriteTable`, and "RETURNING" being the SQL keyword, as in `INSERT INTO ... VALUES ... RETURNING id`? If so: Yeah, I've wished for that as well. Should we consider adding that to `db_merge_into`?

Reply:
Yes, that's what I meant.  Would be good if @ashiklom chimed in about `db_merge_into` since I haven't used that before and I don't completely follow what it's doing internally. But yes, if there's a way to add `RETURNING id` 

In the mean time, what do you want @LiamBurke24 to do immediately? Fields generated via pull-down menus and other queries should be safe (e.g. mimetype ID), but is there a way to properly sanitize anything that's coming in via a text field (e.g. format name)?

Reply:
So the point of `db_merge_into` is to try to solve somewhat elegantly what I felt was the very common problem of updating an SQL table with a local table based on one or more "key" columns. The objective is to have a function that, if one were to run it twice with the same arguments, would _maybe_ do something the first time (if the target SQL table was missing rows compared to the local input, those rows would be added) but would basically _always_ do nothing the second time (because the table has already been updated). I was hoping this would allow us to simplify code for adding various things to the database by eliminating the need for logic like the following (which is, effectively, what `db_merge_into` does using `dplyr` filtering joins:

```
1. Get SQL table IDs
2. Subset local table to IDs not in SQL table
3. If local table has any remaining rows after subset, write them to SQL
```

For an example of this in action, see the [unit test](https://github.com/PecanProject/pecan/blob/develop/base/db/tests/testthat/test.insert.R).

The `db_merge_into` function uses [`insert_table`](https://github.com/PecanProject/pecan/blob/develop/base/db/R/insert_table.R) as a backend, which was designed to be a more robust and user-friendly version of `DBI::dbWriteTable` that also does things like ensure that column types match and automatically subset and align input data to columns that are in the target SQL table.

I was not aware of `RETURNING` as an SQL statement when I came up with this -- that's good to know about for the future. We _could_ add it to `db_merge_into`, but I'm not sure it's necessary. `db_merge_into` currently returns a `dplyr` "lazy SQL query" `data.frame` (which can easily be loaded in memory with just a `dplyr::collect`) which includes the full target SQL table (including ID columns) inner-joined with the input data. So the IDs and any other columns of interest from the SQL table are already in the output.

Reply:
> We don't yet have a fully formed standard method (hence the open status of #395), but in general instead of building an SQL string inside of any PEcAn function we want to assemble the data that we want to insert and then let an existing function to the translation to SQL.

I _strongly_ agree with this. More generally, creating a well-documented API for inserting data into bety should be a top priority for our next releases, as it could make a _lot_ of tasks much easier and safer.

Reply:
@LiamBurke24 you want to try revising to use `db_merge_into`? Sound like this would also eliminate the need for a loop in the formats_variables insert.

Reply:
@LiamBurke24 I tried to make `db_merge_into` easy to use, but ping me on Slack if you have questions. The general idea is to first construct a `data.frame` that looks something like your target SQL table (in this case, `formats_variables`) in terms of column names, and then specify which columns determine the uniqueness of the join via a character vector `by` argument. E.g.

```r
input_formats <- data.frame(
  header = header,
  skip = skip,
  mimetype_id = mimetype_id,
  notes = format_notes,
  name = format_name
)
inserted_data <- db_merge_into(input_formats, "formats_variables", con, by = c("format_name", "mimetype_id"))
```

The `by` argument is optional -- if omitted, it joins based on all columns the two data frames have in common (i.e. if _any_ of the common columns are different, then it assumes you meant to insert a new record). In practice, you usually only want to check some of the columns -- e.g. above, if something with the same `format_name` and `mimetype_id` already exists, then you probably don't want to add a new format with the same `name` and `id` but, say, slightly different `notes`.

Reply:
Never mind. This question is addressed below. Sorry.